### PR TITLE
Issue 1140 twitter

### LINKED ISSUE #1140: Social Provider - Twitter Support
Support Social Login using Twitter in the JSDK

### PR DESCRIPTION
**Pull request by rbhojan**
fixed #1140

### CODE REVIEW

File: api/pom.xml [hunk 1]
```diff
@@ -21,7 +21,7 @@
     <parent>
         <groupId>com.stormpath.sdk</groupId>
         <artifactId>stormpath-sdk-root</artifactId>
-        <version>1.2.2-SNAPSHOT</version>
+        <version>1.2.3-SNAPSHOT</version>
```
Comment (dogeared) [inline api/pom.xml]: These should all be `1.3.0` now.

File: api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java]: 2016. Also, this should be in all new files

File: api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import com.stormpath.sdk.lang.Classes;
+
+/**
+ * Static utility/helper methods for working with userInfo from Social Provider related resources.
+ * Most methods are <a href="http://en.wikipedia.org/wiki/Factory_method_pattern">factory method</a>s used for
+ * creating {@link UserInfoMappingRule} and {@link UserInfoMappingRules} entities.
+ *
+ * @since 1.3.0
+ */
+public final class SocialUserInfoMappingRules {
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java]: This looks very close to what we've done for `SamlAttributeStatementMappingRules`. Is there no opportunity for DRY'ing out this code? IntelliJ even complains about the duplication:

image

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRule.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * <h5>Example</h5>
+ * <p>For example, assume that a mapping rule's {@code name} is {@code foo}, and the rule's
+ * {@link #getAccountAttributes() accountAttributes} collection contains the Stormpath Account field names
+ * of {@code givenName} and {@code surname}.</p>
+ * <p>If an Attribute Statement named {@code foo} is encountered on login to have a value of {@code bar}, then
+ * {@code bar} will automatically be copied to the specified Stormpath Account fields for that user.  This user would
+ * then automatically have a {@code givenName} of {@code bar} and a {@code surname} of {@code bar}.</p>
+ *
+ * @since 1.0.RC8
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRule.java]: copy and paste monster! should be 1.3.0

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRuleBuilder.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRuleBuilder.java]: 2016

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package com.stormpath.sdk.provider.social;
+
+
+import com.stormpath.sdk.resource.Auditable;
+import com.stormpath.sdk.resource.Resource;
+import com.stormpath.sdk.resource.Saveable;
+
+import java.util.List;
+
+public interface UserInfoMappingRules extends Resource, Saveable, Auditable {
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java]: should be a javadoc above interface that has 1.3.0 and info on this interface

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    /**
+     * Specifies the Set of all {@link UserInfoMappingRule}s that indicate how SAML Attribute Statements should
+     * populate one or more Stormpath Account field values after a successful SAML login.
+     *
+     * @param userInfoMappingRules the set of {@link UserInfoMappingRule userInfoMappingRules}s to build a Social provider.
+     *
+     * @return this instance for method chaining.
+     * @since 1.3.0
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java]: not needed here if put above interface declaration

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    void setItems(List<UserInfoMappingRule> userInfoMappingRules);
+
+    /**
+     * Returns the Set of all {@link UserInfoMappingRule}s that indicate how SAML Attribute Statements should
+     * populate one or more Stormpath Account field values after a successful SAML login.
+     *
+     * @return the Set of all {@link UserInfoMappingRule}s that indicate how SAML Attribute Statements should
+     * populate one or more Stormpath Account field values after a successful SAML login.
+     * @since 1.3.0
```
Comment (dogeared) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java]: not needed here if put above interface declaration

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy [hunk 9]
```diff
@@ -39,14 +40,7 @@ import org.testng.annotations.Test
...
 
-import static org.testng.Assert.assertEquals
-import static org.testng.Assert.assertFalse
-import static org.testng.Assert.assertNotEquals
-import static org.testng.Assert.assertNotNull
-import static org.testng.Assert.assertNotSame
-import static org.testng.Assert.assertNull
-import static org.testng.Assert.assertTrue
-import static org.testng.Assert.fail
+import static org.testng.Assert.*
```
Comment (dogeared) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy]: not a requirement in tests, but we prefer the exploded imports.

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultFacebookCreateProviderRequestBuilder.java [hunk 10]
```diff
@@ -35,6 +35,9 @@ protected CreateProviderRequest doBuild(Map<String, Object> map) {
         DefaultFacebookProvider provider = new DefaultFacebookProvider(null, map);
         provider.setClientId(super.clientId);
         provider.setClientSecret(super.clientSecret);
+        if(super.userInfoMappingRules != null) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultFacebookCreateProviderRequestBuilder.java]: space after "if": `if (...`

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGoogleCreateProviderRequestBuilder.java [hunk 11]
```diff
@@ -48,7 +48,9 @@ protected CreateProviderRequest doBuild(Map<String, Object> map) {
         provider.setClientId(super.clientId);
         provider.setClientSecret(super.clientSecret);
         provider.setRedirectUri(this.redirectUri);
-
+        if(super.userInfoMappingRules != null) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGoogleCreateProviderRequestBuilder.java]: space after if

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGithubCreateProviderRequestBuilder.java [hunk 12]
```diff
@@ -35,7 +35,9 @@ protected CreateProviderRequest doBuild(Map<String, Object> map) {
         DefaultGithubProvider provider = new DefaultGithubProvider(null, map);
         provider.setClientId(super.clientId);
         provider.setClientSecret(super.clientSecret);
-
+        if(super.userInfoMappingRules != null) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGithubCreateProviderRequestBuilder.java]: space after if

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultLinkedInCreateProviderRequestBuilder.java [hunk 13]
```diff
@@ -49,6 +49,9 @@ protected CreateProviderRequest doBuild(Map<String, Object> map) {
             provider.setRedirectUri(redirectUri);
         }
 
+        if(super.userInfoMappingRules != null) {
```
Comment (dogeared) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultLinkedInCreateProviderRequestBuilder.java]: space after if

File: api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java (hunk 3 above)
Comment (rbhojan) [inline api/src/main/java/com/stormpath/sdk/provider/social/SocialUserInfoMappingRules.java]: I will address this in the other PR : #1147

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRule.java (hunk 4 above)
Comment (rbhojan) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRule.java]: I will address this in the other PR : #1147

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRuleBuilder.java (hunk 5 above)
Comment (rbhojan) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRuleBuilder.java]: I will address this in the other PR : #1147

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java (hunk 6 above)
Comment (rbhojan) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java]: I will address this in the other PR : #1147

File: api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java (hunk 7 above)
Comment (rbhojan) [inline api/src/main/java/com/stormpath/sdk/provider/social/UserInfoMappingRules.java]: Ok. I will address this in the other PR : #1147

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultLinkedInCreateProviderRequestBuilder.java (hunk 13 above)
Comment (rbhojan) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultLinkedInCreateProviderRequestBuilder.java]: I will address this in the other PR : #1147

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGithubCreateProviderRequestBuilder.java (hunk 12 above)
Comment (rbhojan) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGithubCreateProviderRequestBuilder.java]: I will address this in the other PR : #1147

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGoogleCreateProviderRequestBuilder.java (hunk 11 above)
Comment (rbhojan) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultGoogleCreateProviderRequestBuilder.java]: I will address this in the other PR : #1147

File: impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultFacebookCreateProviderRequestBuilder.java (hunk 10 above)
Comment (rbhojan) [inline impl/src/main/java/com/stormpath/sdk/impl/provider/DefaultFacebookCreateProviderRequestBuilder.java]: I will address this in the other PR : #1147

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy (hunk 9 above)
Comment (rbhojan) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/client/DirectoryIT.groovy]: I will address this in the other PR : #1147

**Comment by rbhojan** [comment]
Fixed all the review comments @dogeared . Please see if this is good to go and approve the PR.

**Comment by dogeared** [comment]
LGTM. Please merge.

**Pull request merged**